src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,144 @@
+            System.arraycopy(tempArray, 0, elements, size, tempArray.length);
+            size = newCapacity;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " of bounds exception");
+        } else {
+            return elements[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " of bounds exception");
+        } else {
+            elements[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " of bounds exception");
+        } else {
+            int newSize = size - 1;
+            final T oldValue = get(index);
+            if (newSize > index) {
+                System.arraycopy(elements, index + 1, elements, index, newSize - index);
+            }
+            size = newSize;
+            elements[newSize] = null;
+            return oldValue;
+        }
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = 0;
+        found:
+        {
+            if (element == null) {
+                for (; index < size; index++) {
+                    if (elements[index] == null) {
+                        remove(index);
+                        break found;
+                    }
+                }
+            } else {
+                for (; index < size; index++) {
+                    if (element.equals(elements[index])) {
+                        remove(index);
+                        break found;
+                    }
+                }
+                throw new NoSuchElementException();
+            }
+        }
+        return element;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        boolean empty = true;
+        for (T element : elements) {
+            if (element != null) {
+                empty = false;
+                break;
+            }
+        }
+        return empty;
+    }
+
+    public static Object[] listToArray(List<?> list) {
+        if (!list.isEmpty()) {
+            Object[] tempArray = new Object[list.size()];
+            for (int index = 0; index < list.size(); index++) {
+                tempArray[index] = list.get(index);
+            }
+            return tempArray;
+        }
+        return null;
+    }
+
+    private void grow(int minimalCapacity) {
+        int oldCapacity = elements.length;
+        double multiplicator = 1;
+        if (oldCapacity == size) {
+            multiplicator = 1.5;
+        }
+        if (oldCapacity > 0 || isEmpty()) {
```

Comment:
why do we need this check?

Reply:
because it should have the default capacity, the internal array should grow 1.5 times when it is full